src/tools/crossgen2/Common/TypeSystem/Interop/IL/MarshalHelpers.cs

```diff
@@ -0,0 +1,693 @@
+
+                case MarshallerKind.AnsiChar:
+                    return context.GetWellKnownType(WellKnownType.Byte);
+
+                case MarshallerKind.FunctionPointer:
+                    return context.GetWellKnownType(WellKnownType.IntPtr);
+
+#if !READYTORUN
+                case MarshallerKind.ByValUnicodeString:
+                case MarshallerKind.ByValAnsiString:
+                    {
+                        var inlineArrayCandidate = GetInlineArrayCandidate(context.GetWellKnownType(WellKnownType.Char), elementMarshallerKind, interopStateManager, marshalAs);
+                        return interopStateManager.GetInlineArrayType(inlineArrayCandidate);
+                    }
+
+                case MarshallerKind.ByValAnsiCharArray:
+                case MarshallerKind.ByValArray:
+                    {
+                        ArrayType arrayType = type as ArrayType;
+                        Debug.Assert(arrayType != null, "Expecting array");
+
+                        var inlineArrayCandidate = GetInlineArrayCandidate(arrayType.ElementType, elementMarshallerKind, interopStateManager, marshalAs);
+
+                        return interopStateManager.GetInlineArrayType(inlineArrayCandidate);
+                    }
+#endif
+
+                case MarshallerKind.LayoutClassPtr:
+                case MarshallerKind.AsAnyA:
+                case MarshallerKind.AsAnyW:
+                    return context.GetWellKnownType(WellKnownType.IntPtr);
+
+                case MarshallerKind.Unknown:
+                default:
+                    throw new NotSupportedException();
+            }
+        }
+
+        internal static MarshallerKind GetMarshallerKind(
+             TypeDesc type,
+             MarshalAsDescriptor marshalAs,
+             bool isReturn,
+             bool isAnsi,
+             MarshallerType marshallerType,
+             out MarshallerKind elementMarshallerKind)
+        {
+            if (type.IsByRef)
+            {
+                type = type.GetParameterType();
+            }
+            TypeSystemContext context = type.Context;
+            NativeTypeKind nativeType = NativeTypeKind.Invalid;
+            bool isField = marshallerType == MarshallerType.Field;
+
+            if (marshalAs != null)
+                nativeType = (NativeTypeKind)marshalAs.Type;
+
+
+            elementMarshallerKind = MarshallerKind.Invalid;
+
+            //
+            // Determine MarshalerKind
+            //
+            // This mostly resembles desktop CLR and .NET Native code as we need to match their behavior
+            // 
+            if (type.IsPrimitive)
+            {
+                switch (type.Category)
+                {
+                    case TypeFlags.Void:
+                        return MarshallerKind.VoidReturn;
+
+                    case TypeFlags.Boolean:
+                        switch (nativeType)
+                        {
+                            case NativeTypeKind.Invalid:
```

Comment:
We start with `NativeTypeKind nativeType = NativeTypeKind.Invalid;` and this is left if MarshalAs is unspecified.

CoreCLR has this "nicer" because they have a `#define` that defines `NATIVE_TYPE_MAX` (which is "our `Invalid`" to `NATIVE_TYPE_DEFAULT`. But both seem correct.